src/Streamly/Internal/Data/Parser/ParserD/Types.hs

```diff
@@ -793,60 +786,59 @@ instance MonadCatch m => Alternative (Parser m a) where
     some = splitSome toList
 
 {-# ANN type ConcatParseState Fuse #-}
-data ConcatParseState sl p = ConcatParseL sl | ConcatParseR p
+data ConcatParseState sl m a b =
+      ConcatParseL sl
+    | forall s. ConcatParseR (s -> a -> m (Step s b)) s (s -> m b)
 
 -- | See 'Streamly.Internal.Data.Parser.concatMap'.
 --
 -- /Internal/
 --
 {-# INLINE concatMap #-}
-concatMap :: MonadThrow m => (b -> Parser m a c) -> Parser m a b -> Parser m a c
+concatMap :: MonadThrow m =>
+    (b -> Parser m a c) -> Parser m a b -> Parser m a c
 concatMap func (Parser stepL initialL extractL) = Parser step initial extract
 
     where
 
-    {-# INLINE scrutInit #-}
-    scrutInit (Parser stepR initialR extractR) = do
+    {-# INLINE initializeR #-}
+    initializeR (Parser stepR initialR extractR) = do
         resR <- initialR
-        case resR of
-            IPartial sr -> do
-                let prsr = Parser stepR (return (IPartial sr)) extractR
-                return $ IPartial $ ConcatParseR prsr
-            IDone br -> return $ IDone br
-            IError err -> return $ IError err
+        return $ case resR of
+            IPartial sr -> IPartial $ ConcatParseR stepR sr extractR
+            IDone br -> IDone br
+            IError err -> IError err
 
     initial = do
         res <- initialL
         case res of
             IPartial s -> return $ IPartial $ ConcatParseL s
-            IDone b -> scrutInit (func b)
+            IDone b -> initializeR (func b)
             IError err -> return $ IError err
 
+    {-# INLINE initializeRL #-}
+    initializeRL n (Parser stepR initialR extractR) = do
```

Comment:
We can probably add `next` `done` and `err` as parameters to this and only keep one of `initializeR` to `initializeRL`. I'll make this change and check perf.

Reply:
I had thought about it but did not do it. can be done.